File: lang/src/main/java/org/xvm/lsp/model/Diagnostic.java
package org.xvm.lsp.model;

import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

/**
 * Immutable diagnostic (error, warning, hint).
 */
public record Diagnostic(
        @NonNull Location location,
        @NonNull Severity severity,
        @NonNull String message,
        @Nullable String code,
        @Nullable String source
) {
    public enum Severity {
        ERROR,
        WARNING,
        INFORMATION,
        HINT
    }

    public static Diagnostic error(
            final @NonNull Location location,
            final @NonNull String message) {
        return new Diagnostic(location, Severity.ERROR, message, null, "xtc");
    }

    public static Diagnostic warning(
            final @NonNull Location location,
            final @NonNull String message) {
        return new Diagnostic(location, Severity.WARNING, message, null, "xtc");
    }

    public static Diagnostic info(
            final @NonNull Location location,
            final @NonNull String message) {
        return new Diagnostic(location, Severity.INFORMATION, message, null, "xtc");
    }
}

File: lang/lsp-server/src/main/kotlin/org/xvm/lsp/model/Diagnostic.kt
package org.xvm.lsp.model

/**
 * Immutable diagnostic (error, warning, hint).
 */
data class Diagnostic(
    val location: Location,
    val severity: Severity,
    val message: String,
    val code: String? = null,
    val source: String? = null,
) {
    enum class Severity {
        ERROR,
        WARNING,
        INFORMATION,
        HINT,
    }

    companion object {
        fun error(
            location: Location,
            message: String,
        ): Diagnostic = Diagnostic(location, Severity.ERROR, message, null, "xtc")

        fun warning(
            location: Location,
            message: String,
        ): Diagnostic = Diagnostic(location, Severity.WARNING, message, null, "xtc")

        fun info(
            location: Location,
            message: String,
        ): Diagnostic = Diagnostic(location, Severity.INFORMATION, message, null, "xtc")
    }
}
